### ISSUE TITLE
Adjust `_tzinfo` implementation to return _DatetimeWithFold or equivalent.

### ISSUE THREAD

**Issue by pganssle** [dateutil/dateutil#286]
Moving this discussion from #57.

Per the discussion there, I am convinced that @abalkin's PEP-495-consistent implementation of ambiguous time handling [in this mini-project](https://github.com/abalkin/tz) is the right way to do things if we want to be consistent across versions when PEP-495 goes into effect.

The short form is that we'd add a class like this:

``` python
    class _DatetimeWithFold(datetime):
        __slots__ = ()

        @property
        def fold(self):
            return 1
```

And rather than spawning new instances of `_tzinfo` to handle fold behavior, we'd handle fold behavior with `fold = getattr(dt, fold, 0)`.

PEP-495 has some (to me) counter-intuitive choices, so we should probably make an effort to thoroughly document timezone behavior.

**Comment by abalkin** [comment]
Let me add a link to the [relevant datetime-sig thread](https://mail.python.org/pipermail/datetime-sig/2015-September/000860.html).  Note that the issue of PEP 495 gap/fold rules being the opposite of the "legacy" rules was discussed in that thread.  Here is Tim's recommendation for how to deal with that in a backward compatible manner:

>  I'd suggest leaving the current implementation alone,
and _adding_ a new wrapping of tzfiles based on Alex's code.
dateutil's get-a-zone factory functions would need to grow some way to
spell "I want a pre-495 tzinfo" or "I want a post-495 tzinfo".  New
functions, optional function flags, global setting ... whatever you
think works best.

See [Tim's post](https://mail.python.org/pipermail/datetime-sig/2015-September/000863.html).

**Comment by pganssle** [comment]
I'm not all *that* worried about changing the legacy behavior. It seems like a lot of needless complexity to maintain two versions of all `tzinfo` implementations just so we can have backwards compatibility around ambiguous dates, considering that anyone who cares about handling ambiguous dates was probably using `pytz` anyway, since up until recently we've had no way to represent them.

I dunno, I'll see if there's an easy way to do it, but in some sense I'd rather just say, "This is the new behavior, going forward."

Another, probably less desirable, option is to use `getattr(dt, fold, 1)` (or possibly `getattr(dt, fold, 0)`) instead of `getattr(dt, fold 0)`. This would make the use of a python with a PEP-495 implementation the clear demarcation line where the behavior changes. There's a case to be made there, but I think it's a bit implicit for my taste.

**Comment by pganssle** [comment]
@abalkin I'm trying to make an implementation here that has the post-495 rules, and I'm noticing that `fromutc` seems to get the wrong value if I have the DST value for "unspecified" default to DST instead of STD, and I'm fairly certain it's because `fromutc` is assuming that ambiguous dates return 0 for DST.

This is because in [CPython's `fromutc`](https://github.com/python/cpython/blob/master/Lib/datetime.py#L976), it basically creates a datetime that is offset by 1 DST offset on the DST side if the correct output is on the DST side, and a datetime that is the correct value but _ambiguous_ if the correct output is on the STD side. If my code interprets unspecified ambiguous dates as being on the DST side, `fromutc()` will then be off by 1 hour:

(using my current working directory, with ambiguous dates defaulting to the DST side)
```python
>>> from datetime import datetime, timedelta
>>> from dateutil.tz import tzlocal, tzutc
>>> EST = tzlocal()
>>> UTC = tzutc()
>>> dt_edt_utc = datetime(2016, 11, 6, 5, 30, tzinfo=UTC)   # Unambiguously EDT
>>> dt_edt = dt_edt_utc.astimezone(EST)
>>> dt_edt, dt_edt.tzname()
(datetime.datetime(2016, 11, 6, 1, 30, tzinfo=tzlocal()), 'EDT')
>>> dt_est = (dt_edt_utc + timedelta(hours=1)).astimezone(EST)   # Should be 1:30 EST
>>> dt_est
datetime.datetime(2016, 11, 6, 2, 30, tzinfo=tzlocal())
```

Since it seems you've tackled this before, do you have any recommendation here?

**Comment by abalkin** [comment]
You cannot reuse the default fromutc - your tzinfo implementation must supply its own.  This is explained in the [Guidelines for New tzinfo Implementations](https://www.python.org/dev/peps/pep-0495/#guidelines-for-new-tzinfo-implementations) section of [PEP 495](https://www.python.org/dev/peps/pep-0495).  You can also see [ZoneInfo.fromutc](https://github.com/abalkin/tz/blob/tz-1a1/tz/zoneinfo.py#L112) in my reference implementation.

**Comment by pganssle** [comment]
@abalkin Hm.. I guess that's what I've done, but it seems wholly unnecessary. If [line 997 of datetime.py](https://github.com/python/cpython/blob/be8de928e5d2f1cd4d4c9c3e6545b170f2b02f1b/Lib/datetime.py#L997) were changed to:

```python
dtdst = dt.replace(fold=1).dst()
```

Then all compliant implementations (with and without variable DST) would work as expected. What is the reason for not doing this?

**Comment by pganssle** [comment]
@abalkin To be clear, I'm just saying that `fromutc()` goes from returning something meaningful (the correct date and time, with no indication of what side of the fold it's on) to something useless (an incorrect date and time) in the case where you're on the STD side of the divide. That change would restore the original behavior.

For most of the `tzinfo` implementations I'm providing (`tzrange`, `tzwin`, etc), there's no problem with an invariant standard time offset, so I'd prefer to fall back on the default algorithm in those cases. 

Another advantage with using the standard algorithm as a starting point is that it all the type checking and such that's done in `fromutc` is done at that level, and I don't have to keep track of handling that.

All that said, it's not a huge deal either way.

**Comment by pganssle** [comment]
@abalkin Moving this discussion over to http://bugs.python.org/issue28602 so we're not having it in two places.

**Comment by pganssle** [comment]
Implemented in #302.
